c10/cuda/CUDAAssert.h

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include <cstdint>
+#include <mutex>
+
+namespace c10 {
+namespace cuda {
+
+constexpr size_t C10_ASSERT_BUFFER_SIZE = 512;
+constexpr size_t C10_ASSERT_ARG_ALIGN_SIZE = sizeof(double);
+
+// Global CUDA kernel assert and error codes
+
+namespace AssertTag {
```

Comment:
I'm a little confused about the function of the AssertTag here. In particular, what I was kind of expecting when we discussed coming up with error tags was that there'd just be a global pool of common errors (like division by zero, index out of bounds) and also tags for files (so we could print filename.cu:lineno and uniquely identify the error site), but not an error tag for every specific call site. You don't really make use these codes right now (except for the multinomial errors; but honestly we can just toss them all in one big enum.) Did you have other ideas about what you were going to use the specific individual codes would be? :)

Reply:
All error-codes that I specified are referenced at the asserts but only a few are used in the format table. I wanted to ensure that every single assert line is identifiable when the assert is printed or the exception analyzed. It is true that this might also be possible by using the line-number and keeping a table of file-name-identifiers and passing the identifier to the assert. My idea was: Since we cannot specify the format string directly at the C10_ASSERT() I would have to keep a table with all the format strings. Currently a lot of C10 asserts have no format string, but ideally more should have one, e.g. in the CPU implementations there are more informative texts for a couple of asserts.

It would be possible to get all the functionality by distributing the information into:

- AssertKind: defines the type of Exception thrown
- FormatStringId: Format String used (optional)
- FileId: Identifies the file name of the assertion